### ISSUE TITLE
[Guide?] How to run v4.5 on google colab

### ISSUE THREAD

**Issue by IbrahimKurz** [lllyasviel/style2paints#151]
![zbiarita](https://user-images.githubusercontent.com/60773442/94641401-cb991000-0313-11eb-9492-cbb40e0a4d51.png)

I hate anything which come along with a fat framework, especially when those extra ~600mb garbage is just for an unnecessary purpose.
So I made some tweaks to make v4.5 run on google colab. If you also want to avoid this totally unoptimized thing cooking your pc, you might be interested.

Unfortunately, I can't provide you guys full code cuz the author of this is clearly willing to keep something in "secret" by obfuscating his codes. ( scriptkiddie-tier dumb obfuscation, tho )

Here are the steps:

- 1. Fork the v4.5 gdrive release to your own gdrive.

- 2. Make a colab notebook, mount gdrive, insert 1st code block like this:

```py
import os
!unzip '/content/drive/My Drive/<your path>/Style2PaintsV45配布/style2paints45beta1214B.zip' > 00.txt
os.rename('/content/style2paints45beta1214B/assets/nets','/content/nets')
os.rename('/content/style2paints45beta1214B/assets/game','/content/game')
os.rename('/content/style2paints45beta1214B/assets/gpu.conf','/content/gpu.conf')
```

run it.

- 3. Modify /content/game/src/project.js, remove all 'window.server_url+' string in it, or just replace it with the attachment.
[project.txt](https://github.com/lllyasviel/style2paints/files/5303064/project.txt)

- 4. Get the decompiled v4.5 code by some mean.
As you see a decompiled script starts with 'def main_exec():', and full of sh\*t like OO0O000OO000OOO00(op), that's it.
Paste the decompiled code to 2nd code block.

- 5. Erase all the import xxx, from xxx import xxxx in the 2nd code,
replace all '8233' with 'viewppp',
paste
```py
%tensorflow_version 1.x
!pip install bottle
import sys
import socket, os
import tensorflow as tf
import numpy as np, datetime, pickle, base64, json, gzip, re, os, shutil
import cv2
from tensorflow.python.client import device_lib
from threading import Thread
from google.colab.output import eval_js
from tqdm import tqdm
from cv2.ximgproc import l0Smooth, createGuidedFilter, guidedFilter
from bottle import route, run, static_file, request, BaseRequest

BaseRequest.MEMFILE_MAX = 10000000


def replaceservname(servaddr):
  data=[]
  with open('/content/game/src/project.js','rt') as fjs:
    data=fjs.read()
    data=data.replace('http://127.0.0.1:8233',servaddr)
  with open('/content/game/src/project.js','wt') as fjs:
    fjs.write(data)
   
def main_exec(viewppp):
    ruul = eval_js( 'google.colab.kernel.proxyPort(' + str(viewppp) +')')[:-1]
    print(ruul+'/index.html')
    replaceservname(ruul)
```
to 2nd code's head (overwrite first line of def main_exec():)

- 6. Comment out these lines:
```py
os.system('start ./dotnet/Style2Paints.exe')
```
```py
Thread(target=O0000O0OO0O0000O0, args=('Now Loading ...', )).start()
```
```py

        if os.path.exists('_pywrap_tensorflow_internal_gpu.pyd'):
            if os.path.exists('_pywrap_tensorflow_internal.pyd'):
                shutil.move('_pywrap_tensorflow_internal.pyd', '_pywrap_tensorflow_internal_cpu.pyd')
            shutil.move('_pywrap_tensorflow_internal_gpu.pyd', '_pywrap_tensorflow_internal.pyd')
        else:
            print('Style2Paints Customed Tensorflow Shifting to CPU.')
            if os.path.exists('_pywrap_tensorflow_internal_cpu.pyd'):
                if os.path.exists('_pywrap_tensorflow_internal.pyd'):
                    shutil.move('_pywrap_tensorflow_internal.pyd', '_pywrap_tensorflow_internal_gpu.pyd')
                shutil.move('_pywrap_tensorflow_internal_cpu.pyd', '_pywrap_tensorflow_internal.pyd')
                
```
- 7. Run 2nd code block, make 3rd code block with
```py
main_exec(8233)
```

8. Run 3rd code block,
after message like
```py
Listening on http://0.0.0.0:8233/
Hit Ctrl-C to quit.
```
shows, click the first url in output which looks like:
https://xxxxxxxxxxxxxxxxxxxxx-8233-colab.googleusercontent.com/index.html
than that's it.

**Comment by SylveonBottle** [comment]
How does this compare to just running the orriginal on your own machine, in terms of performance / time required to color?

**Comment by keng0o** [comment]
How Get decompiled v4.5 code?
@IbrahimKurz

**Comment by 03stevensmi** [comment]
Yes! I would also like to know how to get the decompiled 4.5 code!
Anything helpful on that?

**Comment by EuphoricPenguin** [comment]
I never thought I'd see the day when people were asking for decompiled source in a GitHub issue thread. The repository maintainers need to seriously re-consider this pseudo-FOSS approach if they want to keep the phased up for much longer. Contradictions in the README that clash with the LICENSE, zero source code for a sketchy .exe hosted outside of the releases section on a random personal Google Drive; the list goes on. I'll happily support this project and share it around if we can first sort out the basic issue of sharing code on a website that was created with the intention of facilitating the sharing of code.

**Comment by h4rk8s** [comment]
How Get decompiled v4.5 code?
@IbrahimKurz

**Comment by 03stevensmi** [comment]
@IbrahimKurz yes, I also would like to know how to get the decompiled code!
Could you please find some way to distribute it so everyone can enjoy.

**Comment by basemdabbour** [comment]
Hello, can someone share google Colab for Style2Paints V4.5?.

**Comment by tegaki-tegaki** [comment]
ok, so here's an attempt at an up-to-date colab with "all the batteries"... sorta, the colab follows most of the instructions in the above issue by @IbrahimKurz 

notable changes:
- two years has gone by, colab stopped supporting tensorflow v1, so the original decompiled script has been crammed through `tf_upgrade_v2`!
- the versions of some of the dependencies in the [original github docs](https://github.com/lllyasviel/style2paints/tree/master/V4.5) have been bumped up to "it runs" (with tf v2)
- the original uploaded project has a generous license, so I guess it's fine to decompile, there's detailed instructions on how to decompile it as well in the colab text.

https://colab.research.google.com/drive/12pJbvncZJkIOgfWyNdig37WP4XsP5jVh?usp=sharing

# it fails!?

yes... there's now a logic error, maybe a tensorflow default changed? (from 32 to 56?) who knows! I tried a bit, looked for definitions of both the number `32` as well as `56` in the code, tried to swap to the other and run... etc... essentially "stirring the decompiled soup", and I had to stop and ask myself... why am I trying to debug & make changes to a decompiled python script!? it's 3AM it's time to go to bed! 

# tl;dr 

it's close... and if you know python & tensorflow you could probably change a single line or two and get this to work! ;) (I'm no expert in decompiling (did I get the right file?!), python, tensorflow... etc)

**Comment by 03stevensmi** [comment]
If it helps anyone: 
https://huggingface.co/spaces/HighCWu/Style2Paints-4.5-Radio
https://highcwu-style2paints-4-5-gradio.hf.space/file/ui/web-mobile/index.html

no more viruses or pissing about!
Enjoy!
100% Credit to HighCWu!
Good job man!
